File: retaguarda-software-house/node/retaguarda-sh/src/pdv/pdv-plano-pagamento/pdv-plano-pagamento.service.ts
/*******************************************************************************
Title: T2Ti ERP 3.0                                                                
Description: Service relacionado à tabela [PDV_PLANO_PAGAMENTO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { PdvPlanoPagamento } from './pdv-plano-pagamento.entity';
import { getConnection, QueryRunner } from 'typeorm';
import { Biblioteca } from '../../util/biblioteca';
import { ObjetoPagSeguro } from '../../util/objeto.pagseguro';
import { PdvTipoPlano, Empresa } from '../../entities-export';

@Injectable()
export class PdvPlanoPagamentoService extends TypeOrmCrudService<PdvPlanoPagamento> {

  constructor(
    @InjectRepository(PdvPlanoPagamento) repository) { super(repository); }


	async atualizar(objetoPagSeguroEnviado: ObjetoPagSeguro): Promise<PdvPlanoPagamento> {
		let objetoRetorno: PdvPlanoPagamento;
		const connection = getConnection();
		const queryRunner = connection.createQueryRunner();

		await queryRunner.connect();
		await queryRunner.startTransaction();

		// primeiro verifica se já existe um registro armazenado para a transação, pois neste caso iremos apenas atualizar o registro
		let pdvPlanoPagamentoDB = await connection.manager.findOne(PdvPlanoPagamento, { where: { codigoTransacao: objetoPagSeguroEnviado.codigoTransacao }} );
		if (pdvPlanoPagamentoDB != null) {
			// atualiza o status
			pdvPlanoPagamentoDB.statusPagamento = objetoPagSeguroEnviado.codigoStatusTransacao;

			// se o status for pago, então atualiza a data de pagamento e de expiração do plano
			if (pdvPlanoPagamentoDB.statusPagamento == "3") {
				pdvPlanoPagamentoDB.dataPagamento = new Date();
				pdvPlanoPagamentoDB.dataPlanoExpira = new Date();
				switch (pdvPlanoPagamentoDB.plano) {
					case "M":
						pdvPlanoPagamentoDB.dataPlanoExpira.setMonth(pdvPlanoPagamentoDB.dataPagamento.getMonth() + 1);
						break;
					case "S":
						pdvPlanoPagamentoDB.dataPlanoExpira.setMonth(pdvPlanoPagamentoDB.dataPagamento.getMonth() + 6);
						break;
					case "A":
						pdvPlanoPagamentoDB.dataPlanoExpira.setMonth(pdvPlanoPagamentoDB.dataPagamento.getMonth() + 12);
						break;
					default:
						break;
				}
			}
			objetoRetorno = await queryRunner.manager.save(pdvPlanoPagamentoDB);
			await queryRunner.commitTransaction();
			return objetoRetorno;
		} else {
			// a falta de registro no banco indica que devemos criar um novo registro no banco de dados
			// que só será inserido se tiver vindo um tipo de plano válido na requisição
			let planoContratado = Biblioteca.pegarPlanoPdv(objetoPagSeguroEnviado.descricaoProduto);
			let moduloFiscal = Biblioteca.pegarModuloFiscalPdv(objetoPagSeguroEnviado.descricaoProduto);
			let tipoPlano = await connection.manager.findOne(PdvTipoPlano, { where: { plano: planoContratado, moduloFiscal: moduloFiscal }} );
			if (tipoPlano != null) {
				let pdvPlanoPagamento = new PdvPlanoPagamento({});
				pdvPlanoPagamento.pdvTipoPlano = tipoPlano;
				pdvPlanoPagamento.codigoTransacao = objetoPagSeguroEnviado.codigoTransacao;
				pdvPlanoPagamento.statusPagamento = objetoPagSeguroEnviado.codigoStatusTransacao;
				pdvPlanoPagamento.metodoPagamento = objetoPagSeguroEnviado.metodoPagamento;
				pdvPlanoPagamento.codigoTipoPagamento = objetoPagSeguroEnviado.codigoTipoPagamento;
				pdvPlanoPagamento.valor = objetoPagSeguroEnviado.valorUnitario;
				pdvPlanoPagamento.emailPagamento = objetoPagSeguroEnviado.emailCliente;
				pdvPlanoPagamento.dataSolicitacao = new Date();
				pdvPlanoPagamento.plano = planoContratado;

				// tenta encontrar a empresa pelo e-mail - se não encontrar, o usuário terá
				// que informar o código da transação na App para reconhecer o seu pagamento
				let empresa = await connection.manager.findOne(Empresa, { where: { email: objetoPagSeguroEnviado.emailCliente }} );
				if (empresa != null) {
					pdvPlanoPagamento.empresa = empresa;
				}

				objetoRetorno = await queryRunner.manager.save(pdvPlanoPagamento);
				await queryRunner.commitTransaction();
				return objetoRetorno;
			} else {
				return null;
			}
		}
	}
 
	async consultarPlanoAtivo(cnpj: string): Promise<PdvPlanoPagamento> {
        const connection = getConnection();
        const queryRunner = connection.createQueryRunner();  
        await queryRunner.connect();
        await queryRunner.startTransaction();

        let empresa = await connection.manager.findOne(Empresa, { where: { cnpj: cnpj }} );
        if (empresa != null) {
			let plano = await connection.manager.findOne(PdvPlanoPagamento, { where: { empresa: empresa, dataPlanoExpira: Date.now }} );
			return plano;
		} else {
			return null;
		}
	}	

	async confirmarTransacao(codigoTransacao: string, cnpj: string): Promise<number> {
        const connection = getConnection();
        const queryRunner = connection.createQueryRunner();  
        await queryRunner.connect();
        await queryRunner.startTransaction();

		// remove qualquer hifen que tenha sido colocado pelo usuário
		codigoTransacao = codigoTransacao.replace("-", "");  

		// primeiro verifica se existe o código da transação
		let pdvPlanoPagamento = await connection.manager.findOne(PdvPlanoPagamento, { where: { codigoTransacao: codigoTransacao }} );
		if (pdvPlanoPagamento != null)
		{
			if (pdvPlanoPagamento.empresa != null)
			{
				// achou o código da transação, mas o código já foi utilizado
				return 418;
			}
			else
			{
				// achou o código da transação e não está vinculado a nenhuma empresa
				// vamos vincular o id da empresa e o e-mail de pagamento
				// retorna 200
				let empresa = await connection.manager.findOne(Empresa, { where: { cnpj: cnpj }} );
				empresa.emailPagamento = pdvPlanoPagamento.emailPagamento;
				await queryRunner.manager.save(empresa);
				pdvPlanoPagamento.empresa = empresa;
				await queryRunner.manager.save(pdvPlanoPagamento);
				await queryRunner.commitTransaction();					
				return 200;
			}
		}
		else
		{
			// não achou o código da transação, retorna 404
			return 404;
		}
	}
}

File: retaguarda-software-house/delphi/Service/PDV/PdvPlanoPagamentoService.pas
{*******************************************************************************
Title: T2Ti ERP 3.0
Description: Service relacionado à tabela [PDV_PLANO_PAGAMENTO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************}
unit PdvPlanoPagamentoService;

interface

uses
  PdvPlanoPagamento, ObjetoPagSeguro, PdvTipoPlano, Empresa,
  PdvTipoPlanoService, EmpresaService,
  System.SysUtils, System.Generics.Collections, ServiceBase, StrUtils,
  MVCFramework.DataSet.Utils, Biblioteca;

type

  TPdvPlanoPagamentoService = class(TServiceBase)
  private
  public
    class function ConsultarLista: TObjectList<TPdvPlanoPagamento>;
    class function ConsultarListaFiltro(AWhere: string): TObjectList<TPdvPlanoPagamento>;
    class function ConsultarObjeto(AId: Integer): TPdvPlanoPagamento;
    class function ConsultarObjetoFiltro(AWhere: string): TPdvPlanoPagamento;
    class procedure Inserir(APdvPlanoPagamento: TPdvPlanoPagamento);
    class function Alterar(APdvPlanoPagamento: TPdvPlanoPagamento): Integer;
    class function Excluir(APdvPlanoPagamento: TPdvPlanoPagamento): Integer;
    class function Atualizar(AObjetoPagSeguroEnviado: TObjetoPagSeguro): TPdvPlanoPagamento;
    class function ConsultarPlanoAtivo(ACnpj: string): TPdvPlanoPagamento;
    class function ConfirmarTransacao(ACodigoTransacao: string; ACnpj: string): Integer;
  end;

var
  sql: string;


implementation

{ TPdvPlanoPagamentoService }


class function TPdvPlanoPagamentoService.ConsultarLista: TObjectList<TPdvPlanoPagamento>;
begin
  sql := 'SELECT * FROM PDV_PLANO_PAGAMENTO ORDER BY ID';
  try
    Result := GetQuery(sql).AsObjectList<TPdvPlanoPagamento>;

  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TPdvPlanoPagamentoService.ConsultarListaFiltro(AWhere: string): TObjectList<TPdvPlanoPagamento>;
begin
  sql := 'SELECT * FROM PDV_PLANO_PAGAMENTO where ' + AWhere;
  try
    Result := GetQuery(sql).AsObjectList<TPdvPlanoPagamento>;

  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TPdvPlanoPagamentoService.ConsultarObjeto(AId: Integer): TPdvPlanoPagamento;
begin
  sql := 'SELECT * FROM PDV_PLANO_PAGAMENTO WHERE ID = ' + IntToStr(AId);
  try
    GetQuery(sql);
    if not Query.Eof then
    begin
      Result := Query.AsObject<TPdvPlanoPagamento>;

    end
    else
      Result := nil;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TPdvPlanoPagamentoService.ConsultarObjetoFiltro(AWhere: string): TPdvPlanoPagamento;
begin
  sql := 'SELECT * FROM PDV_PLANO_PAGAMENTO where ' + AWhere;
  try
    GetQuery(sql);
    if not Query.Eof then
    begin
      Result := Query.AsObject<TPdvPlanoPagamento>;

    end
    else
      Result := nil;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TPdvPlanoPagamentoService.ConsultarPlanoAtivo(ACnpj: string): TPdvPlanoPagamento;
var
  Filtro: string;
  Empresa: TEmpresa;
begin
  Result := nil;
  Filtro := 'CNPJ = "' + ACnpj + '"';
  Empresa := TEmpresaService.ConsultarObjetoFiltro(Filtro);
  if Assigned(Empresa) then
  begin
    Filtro := 'ID_EMPRESA = ' + Empresa.Id.ToString + ' AND DATA_PLANO_EXPIRA >= ' + Biblioteca.DateToSQL(now);
    Result := ConsultarObjetoFiltro(Filtro);
  end;
end;

class function TPdvPlanoPagamentoService.ConfirmarTransacao(ACodigoTransacao: string; ACnpj: string): Integer;
var
  Filtro: string;
  Empresa: TEmpresa;
  PdvPlanoPagamento: TPdvPlanoPagamento;
begin
  // remove qualquer hifen que tenha sido colocado pelo usuário
  ACodigoTransacao := StringReplace(ACodigoTransacao, '-', '', [rfReplaceAll, rfIgnoreCase]);
  // primeiro verifica se existe o código da transação
  Filtro := 'CODIGO_TRANSACAO = "' + ACodigoTransacao + '"';
  PdvPlanoPagamento := ConsultarObjetoFiltro(filtro);
  if Assigned(PdvPlanoPagamento) then
  begin
    if PdvPlanoPagamento.IdEmpresa > 0 then
    begin
      // achou o código da transação, mas o código já foi utilizado
      Result := 418;
    end
    else
    begin
      // achou o código da transação e não está vinculado a nenhuma empresa
      // vamos verificar se ela já foi paga - statusPagamento = 3
      if PdvPlanoPagamento.StatusPagamento = '3' then
      begin
        // vamos vincular o id da empresa e o e-mail de pagamento
        // retorna 200
        Filtro := 'CNPJ = "' + ACnpj + '"';
        Empresa := TEmpresaService.ConsultarObjetoFiltro(Filtro);
        Empresa.EmailPagamento := PdvPlanoPagamento.EmailPagamento;
        AlterarBase(Empresa, 'EMPRESA');

        PdvPlanoPagamento.IdEmpresa := Empresa.Id;
        AlterarBase(PdvPlanoPagamento, 'PDV_PLANO_PAGAMENTO');

        Result := 200;
      end
      else
      begin
        Result := 202;
      end;


    end;
  end
  else
  begin
    // não achou o código da transação, retorna 404
    Result := 404;
  end;
end;

class procedure TPdvPlanoPagamentoService.Inserir(APdvPlanoPagamento: TPdvPlanoPagamento);
begin
  APdvPlanoPagamento.ValidarInsercao;
  APdvPlanoPagamento.Id := InserirBase(APdvPlanoPagamento, 'PDV_PLANO_PAGAMENTO');
end;

class function TPdvPlanoPagamentoService.Alterar(APdvPlanoPagamento: TPdvPlanoPagamento): Integer;
begin
  APdvPlanoPagamento.ValidarAlteracao;
  Result := AlterarBase(APdvPlanoPagamento, 'PDV_PLANO_PAGAMENTO');
end;

class function TPdvPlanoPagamentoService.Atualizar(AObjetoPagSeguroEnviado: TObjetoPagSeguro): TPdvPlanoPagamento;
var
  PdvPlanoPagamento, PdvPlanoPagamentoDB: TPdvPlanoPagamento;
  TipoPlano: TPdvTipoPlano;
  Empresa: TEmpresa;
  Filtro, PlanoContratado, ModuloFiscal: string;
begin
  // primeiro verifica se já existe um registro armazenado para a transação, pois neste caso iremos apenas atualizar o registro
  Filtro := 'CODIGO_TRANSACAO = "' + AObjetoPagSeguroEnviado.CodigoTransacao + '"';
  PdvPlanoPagamentoDB := TPdvPlanoPagamentoService.ConsultarObjetoFiltro(Filtro);
  if Assigned(PdvPlanoPagamentoDB) then
  begin
    // atualiza o status
    PdvPlanoPagamentoDB.StatusPagamento := AObjetoPagSeguroEnviado.CodigoStatusTransacao;

    // se o status for pago, então atualiza a data de pagamento e de expiração do plano
    if PdvPlanoPagamentoDB.StatusPagamento = '3' then
    begin
      PdvPlanoPagamentoDB.DataPagamento := now;
      case AnsiIndexStr(PdvPlanoPagamentoDB.Plano, ['M', 'S', 'A']) of
        0 : PdvPlanoPagamentoDB.DataPlanoExpira := now + 30;
        1 : PdvPlanoPagamentoDB.DataPlanoExpira := now + 180;
        2 : PdvPlanoPagamentoDB.DataPlanoExpira := now + 365;
      end;
    end
    else
    begin
      PdvPlanoPagamentoDB.DataPagamento := 0;
      PdvPlanoPagamentoDB.DataPlanoExpira := 0;
    end;

    AlterarBase(PdvPlanoPagamentoDB, 'PDV_PLANO_PAGAMENTO');
    result := PdvPlanoPagamentoDB;
  end
  else
  begin
    // a falta de registro no banco indica que devemos criar um novo registro no banco de dados
    // que só será inserido se tiver vindo um tipo de plano válido na requisição
    PlanoContratado := Biblioteca.PegarPlanoPdv(AObjetoPagSeguroEnviado.DescricaoProduto);
    ModuloFiscal := Biblioteca.PegarModuloFiscalPdv(AObjetoPagSeguroEnviado.DescricaoProduto);
    Filtro := 'PLANO = "' + PlanoContratado + '" AND MODULO_FISCAL = "' + ModuloFiscal + '"';
    TipoPlano := TPdvTipoPlanoService.ConsultarObjetoFiltro(Filtro);
    if Assigned(TipoPlano) then
    begin
      PdvPlanoPagamento := TPdvPlanoPagamento.Create;
      PdvPlanoPagamento.IdPdvTipoPlano := TipoPlano.Id;
      PdvPlanoPagamento.CodigoTransacao := AObjetoPagSeguroEnviado.CodigoTransacao;
      PdvPlanoPagamento.StatusPagamento := AObjetoPagSeguroEnviado.CodigoStatusTransacao;
      PdvPlanoPagamento.MetodoPagamento := AObjetoPagSeguroEnviado.MetodoPagamento;
      PdvPlanoPagamento.CodigoTipoPagamento := AObjetoPagSeguroEnviado.CodigoTipoPagamento;
      PdvPlanoPagamento.Valor := AObjetoPagSeguroEnviado.ValorUnitario;
      PdvPlanoPagamento.emailPagamento := AObjetoPagSeguroEnviado.EmailCliente;
      PdvPlanoPagamento.DataSolicitacao := now;
      PdvPlanoPagamento.Plano := PlanoContratado;

      // se o status for pago, então atualiza a data de pagamento e de expiração do plano
      if AObjetoPagSeguroEnviado.CodigoStatusTransacao = '3' then
      begin
        PdvPlanoPagamento.DataPagamento := now;
        case AnsiIndexStr(PlanoContratado, ['M', 'S', 'A']) of
        0 : PdvPlanoPagamento.DataPlanoExpira := now + 30;
        1 : PdvPlanoPagamento.DataPlanoExpira := now + 180;
        2 : PdvPlanoPagamento.DataPlanoExpira := now + 365;
        end;
      end;

      // tenta encontrar a empresa pelo e-mail - se não encontrar, o usuário terá
      // que informar o código da transação na App para reconhecer o seu pagamento
      Filtro := 'EMAIL = "' + AObjetoPagSeguroEnviado.EmailCliente + '"';
      Empresa := TEmpresaService.ConsultarObjetoFiltro(filtro);
      if Assigned(Empresa) then
      begin
        PdvPlanoPagamento.IdEmpresa := Empresa.Id;
      end;

      InserirBase(PdvPlanoPagamento, 'PDV_PLANO_PAGAMENTO');
      result := PdvPlanoPagamento;
    end
    else
    begin
      result := nil;
    end;
  end;
end;

class function TPdvPlanoPagamentoService.Excluir(APdvPlanoPagamento: TPdvPlanoPagamento): Integer;
begin
  APdvPlanoPagamento.ValidarExclusao;
  
  Result := ExcluirBase(APdvPlanoPagamento.Id, 'PDV_PLANO_PAGAMENTO');
end;


end.
